www/js/con-author.js

```diff
@@ -1234,6 +1237,111 @@ define([
 				}
 			}
 			return {validValue: entityDesc, correctness: correctness};
+		},
+		updateSlotVariables: function(){
+			var slotMap = this.getSchemaProperties("Mapping");
+			var schema = registry.byId(this.controlMap.schemas).get("value");
+			//for each of these combo boxes we have to generate the options	respectively
+			var token = this.generateVariablesForSlots(slotMap);
+			//labels and combo boxes have to be generated in the slots container
+			//destroy any previously generated comboboxes, html
+			var widgets = dijit.findWidgets(dom.byId("variableSlotControlsbox"));
+			dojo.forEach(widgets, function(w){
+				w.destroyRecursive(true);
+			});
+			var slotContHtml = "";
+			for(var p in slotMap){
+				slotContHtml = slotContHtml+'<div class="fieldgroup"><label for="holder'+schema+this.currentID+slotMap[p]+'">'+slotMap[p]+'</label><input id="holder'+schema+this.currentID+slotMap[p]+'" ></div>';
+			}
+			html.set(dom.byId("variableSlotControlsbox"), slotContHtml);
+			var varAr = this._model.getAllVariables();
+			for(var p in slotMap){
+				var curAr = [];
+				var defVar = [{id: ""+p+token, name: ""+p+token}];
+				//concatenate all variables and current default variable for the respective combo box
+				curAr = defVar.concat(varAr);
+				var stateStore = new memory({ data: curAr });
+				var curDiv = 'holder'+schema+this.currentID+slotMap[p];
+				var currentComboBox = new comboBox({
+										store: stateStore,
+										searchAttr: "name",
+										placeholder: "Select your variable Name"   
+										}, curDiv);
+				currentComboBox.startup();
+			}
+		},
+		generateVariablesForSlots: function(slotMap){
+			//This function generates a variable slot map in the session storage if not existing
+			// and returns the appropriate variable number token which will be used as suffix in the dynamic variable comboboxes
+			if(!sessionStorage.getItem("slot_number_map") || sessionStorage.getItem("slot_number_map")!= ""){
+				// Generate the map first time
+				var numberMap = {};
+				var varAr = this._model.getAllVariables();
+				//var varAr = ["D1", "R2", "T3", "D5", "R15", "T4"];
+				for(var i=0; i<varAr.length; i++){
+					//the variables are generally in mostcases of the form D1, id4 etc
+					var nump = varAr[i].match(/\d+$/);
+					var strp = varAr[i].replace(/\d+$/,"");
+					if(!numberMap[nump[0]]){
+						numberMap[nump[0]] = [strp];
+					}
+					else{
+						numberMap[nump[0]].push(strp);
+					}
+				}
+				sessionStorage.setItem("slot_number_map", JSON.stringify(numberMap));
+			}
+			//now we have slot number map, we need to check the map and return appropriate vars
+			var numGenOb = JSON.parse(sessionStorage.getItem("slot_number_map"));
+			var numList = Object.keys(numGenOb);
+			var slotVars = Object.keys(slotMap);
+			var nextNum;
+			if(numList.length < 1){
```

Comment:
The implementation seems overly complex.  Couldn't we just do something like this?

```
var nextSubscript = 1;
var finished = false;
while (!finished) {
	if (nextSubscript in numGenOb){
		// This subscript is in use by some variable, but maybe not one in the current schema
		if (isSubscriptInUse(nextSubscript,slotVars,numGenOb)){
			// This subscript is in use, increment and try again
			nextSubscript++;
			continue;
		} else {
			// The subscript was in the map, but not in use by any of this schema's variables
			finished = true;
			continue;
		}
	} else {
		// The subscript isn't in the map yet
		finished = true;
	}
}
return nextSubscript;
},

/* 
	isSubscriptInUse
	Given a subscript, list of variables, and the numGenOb
	Return true if the subscript is already in use by any of the variables
*/
isSubscriptInUse: function(subscript,slotVars,numGenOb){
	slotVars.some(function(v){
					return numGenOb[subscript].includes(v);
				  }
		}
	}
}

```